Fix non-owner not able to delete their membership

stnguyen90 opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to appwrite here: https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## What does this PR do?

This PR fixes a bug preventing non-owners from deleting their membership

## Test Plan

I verified my changes by:

1. Create Team
2. Add User 1 as Owner
3. Add User 2 as Member
4. Log In with client SDK as User 2
5. Delete Membership

I've also updated the e2e tests to include a test for deleting membership as non-owner.

## Related PRs and Issues

Fixes https://github.com/appwrite/appwrite/issues/2405

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

stnguyen90 commented:
I just checked the db-alpha branch and this seems to still be a problem: https://github.com/appwrite/appwrite/blob/0b934c5e1dc52267eb31277c8251c0d93d496d05/app/controllers/api/teams.php#L797-L799

Meldiron commented:
I got my eyes on this today. It's a valid bud and I replicated the issue to double-check its scope. I'll continue to review this PR by making sure `Authorization::disable` and `Authorization::reset` is a secure approach. So far, great issue&PR, you have done an amazing job! 💪

Meldiron reviewed (changes requested):
In the meantime, the only thing I would improve about this PR is to add `Test for SUCCESS ` and `Test for FAILURE` titles back. We use this to differentiate between tests that expect to work (2XX), and those expecting errors (4XX)

What you did is awesome (explaining specific test), but let's make sure to keep these `SUCCESS` and `FAILURE` titles above that. For example, it could look like:

```php
        /**
         * Test for SUCCESS
         */

        /**
         * Test 1 description
         */
         // Test 1

        /**
         * Test 2 description
         */
         // Test 2
         
        /**
         * Test for FAILURE
         */

         /**
         * Test 3 description
         */
         // Test 3
```

stnguyen90 commented:
So I added the SUCCESS and FAILURE comments back, but I changed the order so that I could add a couple of failure tests. It's easier to do the FAILURE tests first because the memberships would be deleted by the SUCCESS tests. Is this okay?

eldadfux commented on app/controllers/api/teams.php:
```diff
@@ -762,14 +763,20 @@
         }
 
         if ($membership->getAttribute('confirm')) { // Count only confirmed members
+            Authorization::disable();
```
This looks solid. As @Meldiron suspected, we have moved away from `disable` and `reset` methods to use the new `skip` method which is considered safer.

stnguyen90 replied:
This `skip()` function is a great addition, but I couldn't use it because it seems like that that function is in the db-alpha branch rather than master.

eldadfux reviewed (changes requested):
@Meldiron we also need to verify this fix apply to the DB refactor in v0.12.
@stnguyen90 awesome work!

Meldiron commented:
> So I added the `SUCCESS` and `FAILURE` comments back, but I changed the order so that I could add a couple of failure tests. It's easier to do the `FAILURE` tests first because the memberships would be deleted by the `SUCCESS` tests. Is this okay?

That's fine IMO. 

---

As @eldadfux  mentioned, let's try this with `db-alpha` code to make sure it will work with that. Also, this will solve your problem of not having `skip` function. To do that, you can merge `db-alpha` branch into your branch. Then, I would suggest re-running composer update, dropping all data and re-building images:

```
docker run --rm --interactive --tty \
  --volume $PWD:/app \
  composer update --ignore-platform-reqs --optimize-autoloader --no-plugins --no-scripts --prefer-dist
```

```
docker-compose down -v
```

```
docker-compose up -d --build
```

To play around with db-alpha, you will need new SDK. For web, that can be accessed using

```
$ npm install aw-db-alpha --save

import { Appwrite } from "aw-db-alpha";
```

More about DB refactor can be found here: https://warm-tray-285.notion.site/Appwrite-0-12-Alpha-The-Database-Refactor-70f2c1d234394d9ab9d0d548fe5449bc

Meldiron commented:
Took a few mins and can confirm the bug behaves exactly the same on `db-alpha`.

kodumbeats commented:
@Meldiron @stnguyen90 If we want this change in the DB refactor, please target a PR against `feat-database-indexing`, not `db-alpha` 🙏🏻

eldadfux commented:
@stnguyen90 please let me know if you can continue working on this, otherwise, @appwrite/core can pick it up.

stnguyen90 commented:
@eldadfux yes, I'll try to update this.

stnguyen90 commented:
I've fixed the merge conflicts and unit tests work locally. The CI build is failing, but I don't think it's related to my changes O.o

TorstenDittmann commented:
Restarted the CI 👍🏻

eldadfux commented:
Test have completed with this errors:

![image](https://user-images.githubusercontent.com/1297371/149954550-a839eaa9-9403-4e71-b32a-a56b2a536197.png)

Seems like deleting a team returns 500

stnguyen90 commented:
Ci checks passing now!

Meldiron reviewed (approved):
Looks good, great job 💪

eldadfux commented:
@stnguyen90 thank you 🙌
